### ISSUE TITLE
RightCurly: 'alone' mode skips empty blocks

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#3775]
```
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="RightCurly">
            <property name="id" value="RightCurlyAlone"/>
            <property name="option" value="alone"/>
        </module>
        <module name="RightCurly">
            <property name="id" value="RightCurlySame"/>
            <property name="option" value="same"/>
        </module>
    </module>
</module>

$ cat MyClass.java 
public final class MyClass {
    public void foo(String publicPar){
      do { } while (publicPar != null); //OK //expected: ok for SAME, violation for ALONE
      do{
      }while(System.currentTimeMillis() < 0);//OK //expected: ok for SAME, violation for ALONE
      if (true) { } //OK //expected: violation for SAME, violation for ALONE
    }

    // example from http://checkstyle.sourceforge.net/property_types.html#rcurly, in "same".
    public long getId() {return id;} //OK //expected: violation for SAME, violation for ALONE
}

$ java -jar checkstyle-7.3-all.jar -c config.xml MyClass.java
Starting audit...
Audit done.
```

All cases are violations for "same" and "alone". There should be no violation only for "alone_or_singleline" mode.

**Comment by romani** [comment]
@MEZk , please address. It looks lie caused by last update on `alone_or_singleline` .

**Comment by rnveach** [comment]
@romani This looks similar to what I messaged you about.
I think the issue is related to `shouldCheckLastRcurly` at https://github.com/checkstyle/checkstyle/blob/2551676ccf743822951d42658085c70e7aa869bd/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java#L216 .
When we are dealing with last curly of a chain (if-else or try-catch-finally) we ignore all other checks and only make sure the right curly is not on the same line as the next token. If there is no chain, it assumes it is the last of the chain.

**Comment by MEZk** [comment]
@romani 

> All cases are violations for "same" and "alone".

> do { } while (publicPar != null); //OK //expected: ok for SAME, violation for ALONE

It is not a violation for SAME. Right curly is on the same line as a next statement.

**Comment by MEZk** [comment]
@romani 
So final examples:

```java
public class InputRightCurlyEmptyBlocks {
    public void foo(String publicPar) {
        do { } while (publicPar != null); //OK for SAME and ALONE_OR_SINGLELINE, violation for ALONE
        if (true) { } //violation for SAME, violation for ALONE, OK for ALONE_OR_SINGLELINE
    } // OK for ALONE and ALONE_OR_SINGLELINE, violation FOR SAME
} // OK for ALONE and ALONE_OR_SINGLELINE, violation FOR SAME
```

```java
public class InputRightCurlySingleLineBlocks {
    public long getId() {return 1L;} //OK for ALONE_OR_SINGLELINE, violation for SAME, violation for ALONE
} // OK for ALONE and ALONE_OR_SINGLELINE, violation FOR SAME
```

@romani @rnveach 
Please, check this out, before I start fixing the issue.

**Comment by MEZk** [comment]
@romani @rnveach 
We also need to define what ```single line block``` means for the check. From my point of view these are single line blocks:

1) do-while
```java
do { } while (publicPar != null);
```

2) while-do
```java
while (a > 5) { a--; }
```

3) if-else
```java
if (a > 4) { a++; }
```

4) if-else-if
```java
if (a < 2) { --a; } else if (a > 3) { a++: }
```

5) for-loop
```java
for (int i = 1; i < 10; i++) { System.out.println(i); }
```

6) class
```java
class A { private int a; }
```

7) static / instance init block

```java
private int a;
private static int b;
{  a = 2; }
static { b = 3; }
```

**Comment by MEZk** [comment]
@romani 
Can you check the examples?

**Comment by romani** [comment]
>It is not a violation for SAME. Right curly is on the same line as a next statement.

Right.

>Please, check this out, before I start fixing the issue.

Confirmed. Right behavior described.
We need to update xdoc for "same" mode to describe that it will violate all cases that are not for multi locks

**Comment by MEZk** [comment]
@romani  @rnveach 

In the current implementation of brace policy `same` the current block of code is treated as valid:

```java
        if (true) {

        } else { // line 3

        } // line 5
```

However, according to [the documentation](http://checkstyle.sourceforge.net/property_types.html#rcurly) it should not be valid as the documentation states that if brace policy is same then the brace should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else or try/catch/finally). 

If it is a false negative and we should rise a violation on the line 3 then there will be a situation in which we demand that user need to place the next statement after else. I think it is not correct. What is your opinion?

**Comment by MEZk** [comment]
@romani @rnveach 
These are the UTs which I suggest for the check:
https://github.com/MEZk/checkstyle/blob/ff1a26b8a1730317fa90e452654b0fcd9b6089a9/src/test/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheckTest.java#L341-L437
These are the test input files:
[InputRightCurlyEmptyBlocksAlone.java](https://github.com/MEZk/checkstyle/blob/ff1a26b8a1730317fa90e452654b0fcd9b6089a9/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyEmptyBlocksAlone.java#L3)
[InputRightCurlyEmptyBlocksAloneOrSingleline.java](https://github.com/MEZk/checkstyle/blob/ff1a26b8a1730317fa90e452654b0fcd9b6089a9/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyEmptyBlocksAloneOrSingleline.java#L3)
[InputRightCurlyEmptyBlocksSame.java](https://github.com/MEZk/checkstyle/blob/ff1a26b8a1730317fa90e452654b0fcd9b6089a9/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyEmptyBlocksSame.java)
[InputRightCurlySingleLineBlocksAlone.java](https://github.com/MEZk/checkstyle/blob/ff1a26b8a1730317fa90e452654b0fcd9b6089a9/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlySingleLineBlocksAlone.java)
[InputRightCurlySingleLineBlocksAloneOrSingleline.java](https://github.com/MEZk/checkstyle/blob/ff1a26b8a1730317fa90e452654b0fcd9b6089a9/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlySingleLineBlocksAloneOrSingleline.java)
[InputRightCurlySingleLineBlocksSame.java](https://github.com/MEZk/checkstyle/blob/ff1a26b8a1730317fa90e452654b0fcd9b6089a9/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlySingleLineBlocksSame.java)

**Comment by romani** [comment]
> Thus, why line 4 is OK and line 17, 20, 21 are violations? 

as `same` is for multi-blocks 17,20,21 should not be violation.

> Should we also treat right curly brace which closes method or class as a violation if brace policy is set to same and CLASS_DEF and METHOD_DEF are in the set or tokens specified by user?

to my mind there should be no violations, `same` is only for multi-blocks. ONLY!

> You suggested that "if (true) { } "is a violation for same.

`same` should care only multi-blocks cases. This case is very tricky. As it is future IF-ELSE that would demand "}" to be located on next line. So user will need to only add " esle { ...." and do not need to format(move to new line) "}" of IF.

**Comment by MEZk** [comment]
@romani 
Fixing of this issue is postponed till investigation of https://github.com/checkstyle/checkstyle/issues/3921 and https://github.com/checkstyle/checkstyle/issues/3919.

**Comment by MEZk** [comment]
This issue is waiting for https://github.com/checkstyle/checkstyle/issues/4022
https://github.com/checkstyle/checkstyle/issues/4022 should separate the ALONE logic.

**Comment by romani** [comment]
@MEZk , as blocker issue is marked as checkstyle8 , this issue become checkstyle8 too.
I am unassigning you from this issue for now.

**Comment by Vampire** [comment]
@romani by doing regression tests I found that I "accidentally" fixed this issue while working on #6466 which actually is also mentioned in here, the lines 17,20,21 question.

But I have a question to two cases mentioned here to clarify the need for further tweakage.

That empty blocks are skipped is intentional in the current code, as in `shouldBeAloneOnLineWithAloneOption` the logic is `!isAloneOnLine && !isEmptyBody`.
This was introduced 2014 for #213 and was meant to allow
```java
if (true) {}
```
but thereby also allowed things like
```java
try {
    // ignore
} catch (RuntimeException e) {
} finally {
}
```
in alone mode.
To preserve the fix for #213 but not ignore all empty blocks I had changed this logic to be `!isAloneOnLine && (hasLineBreakBefore || !isEmptyBody)`.

So I fixed most of the cases mentioned here, except two.
You wrote `do { } while (publicPar != null)` should be a violation, but with the logic for #213 it is not.
So should the fix for #213 be undone, or is this the expected behavior?
And of course the same for `if (true) { }`.
And for the case #213 is all about.

**Comment by Vampire** [comment]
So now there are two contraditory false-positive statements, #213 and #6345.
Ignoring that fact hence here, this is fixed with #6565.
